cps/kni.c

```diff
@@ -0,0 +1,220 @@
+#include "kni.h"
+
+#define INSMOD_CMD "insmod "
+#define KNI_KMOD "rte_kni"
+#define RMMOD_CMD ("rmmod " KNI_KMOD)
+
+extern long init_module(void *, unsigned long, const char *);
+extern long delete_module(const char *, unsigned int);
+
+int
+kni_change_if(uint8_t port_id, uint8_t if_up)
+{
+	return (if_up)
+		? rte_eth_dev_set_link_up(port_id)
+		: rte_eth_dev_set_link_down(port_id);
+}
+
+int
+kni_change_mtu(uint8_t port_id, unsigned new_mtu)
+{
+	if (unlikely(new_mtu > ETHER_MAX_LEN))
+		return -EINVAL;
+
+	return rte_eth_dev_set_mtu(port_id, new_mtu);
+}
+
+/*
+ * Inserting and removing modules.
+ *
+ * This code is adapted for use in DPDK from the source code of
+ * insmod and rmmod in the module-init-tools package.
+ */
+
+static void *
+grab_file(const char *filename, unsigned long *size)
+{
+	unsigned int max = 16384;
+	int ret, err_save;
+	void *buffer;
+
+	int fd = open(filename, O_RDONLY, 0);
+	if (fd < 0)
+		return NULL;
+
+	buffer = rte_malloc("kni_kmod", max, 0);
+	if (buffer == NULL)
+		goto out_error;
+
+	*size = 0;
+	while ((ret = read(fd, (char *)buffer + *size, max - *size)) > 0) {
+		*size += ret;
+		if (*size == max) {
+			void *p = rte_realloc(buffer, max *= 2, 0);
+			if (p == NULL)
+				goto out_error;
+			buffer = p;
+		}
+	}
+	if (ret < 0)
+		goto out_error;
+
+	close(fd);
+	return buffer;
+
+out_error:
+	err_save = errno;
+	rte_free(buffer);
+	close(fd);
+	errno = err_save;
+	return NULL;
+}
+
+static const char *
+moderror(int err)
+{
+	switch (err) {
+	case ENOEXEC:
+		return "Invalid module format";
+	case ENOENT:
+		return "Unknown symbol in module";
+	case ESRCH:
+		return "Module has wrong symbol version";
+	case EINVAL:
+		return "Invalid parameters";
+	default:
+		return strerror(err);
+	}
+}
+
+int
+lua_init_kni(const char *kni_kmod_path, unsigned int num_kni)
+{
+	unsigned long len;
+	int ret;
+
+	void *file = grab_file(kni_kmod_path, &len);
+	if (file == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "insmod: can't read '%s': %s\n",
+			kni_kmod_path, strerror(errno));
+		return -1;
+	}
+
+	ret = init_module(file, len, "");
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"insmod: error inserting '%s': %d %s\n",
+			kni_kmod_path, ret, moderror(errno));
+		rte_free(file);
+		return ret;
+	}
+	rte_free(file);
+
+	rte_kni_init(num_kni);
+	return 0;
+}
+
+static int
+check_usage(const char *modname)
+{
+	FILE *module_list;
+	char line[10240], name[64];
+	int ret = 0;
+
+	module_list = fopen("/proc/modules", "r");
+	if (module_list == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"Can't open /proc/modules: %s\n", strerror(errno));
+		return -1;
+	}
+
+	while (fgets(line, sizeof(line) - 1, module_list) != NULL) {
+		unsigned long size, refs;
+		int scanned;
+
+		if (strchr(line, '\n') == NULL) {
+			RTE_LOG(ERR, GATEKEEPER, "long line broke rmmod.\n");
+			ret = -1;
+			goto out;
+		}
+
+		scanned = sscanf(line, "%s %lu %lu", name, &size, &refs);
+		if (strcmp(name, modname) != 0)
+			continue;
+
+		if (scanned < 2) {
```

Comment:
We should test that `sscanf()` successfully did its job before comparing the module names. Moreover, making the test `<=` would enable us to move the `if` statement inside of this `if` statement.